airflow/api_connexion/endpoints/task_instance_endpoint.py

```diff
@@ -14,39 +14,234 @@
+        query = query.join(DR, and_(TI.dag_id == DR.dag_id, TI.execution_date == DR.execution_date))
+        query = query.filter(DR.run_id == dag_run_id)
+    query = _apply_range_filter(
+        query, key=DR.execution_date, value_range=(execution_date_gte, execution_date_lte),
+    )
+    query = _apply_range_filter(query, key=TI.start_date, value_range=(start_date_gte, start_date_lte))
+    query = _apply_range_filter(query, key=TI.end_date, value_range=(end_date_gte, end_date_lte))
+    query = _apply_range_filter(query, key=TI.duration, value_range=(duration_gte, duration_lte))
+    query = _apply_array_filter(query, key=TI.state, values=state)
+    query = _apply_array_filter(query, key=TI.pool, values=pool)
+    query = _apply_array_filter(query, key=TI.queue, values=queue)
+
+    query = query.join(
+        SlaMiss,
+        and_(
+            SlaMiss.dag_id == TI.dag_id,
+            SlaMiss.task_id == TI.task_id,
+            SlaMiss.execution_date == TI.execution_date,
+        ),
+        isouter=True,
+    ).add_entity(SlaMiss)
+    task_instances = query.offset(offset).limit(limit).all()
+    total_entries = session.query(
+        func.count(func.distinct(TI.dag_id, TI.execution_date, TI.task_id))
+    ).scalar()
+    return task_instance_collection_schema.dump(
+        TaskInstanceCollection(task_instances=task_instances, total_entries=total_entries)
+    )
 
 
 @security.requires_authentication
-def get_task_instances_batch():
+@provide_session
+def get_task_instances_batch(session=None):
     """
     Get list of task instances.
     """
-    raise NotImplementedError("Not implemented yet.")
+    body = request.get_json()
+    try:
+        data = task_instance_batch_form.load(body)
+    except ValidationError as err:
+        raise BadRequest(detail=str(err.messages))
+    query = session.query(TI)
+
+    query = _apply_array_filter(query, key=TI.dag_id, values=data["dag_ids"])
+    query = _apply_range_filter(
+        query, key=TI.execution_date, value_range=(data["execution_date_gte"], data["execution_date_lte"]),
+    )
+    query = _apply_range_filter(
+        query, key=TI.start_date, value_range=(data["start_date_gte"], data["start_date_lte"]),
+    )
+    query = _apply_range_filter(
+        query, key=TI.end_date, value_range=(data["end_date_gte"], data["end_date_lte"])
+    )
+    query = _apply_range_filter(
+        query, key=TI.duration, value_range=(data["duration_gte"], data["duration_lte"])
+    )
+    query = _apply_array_filter(query, key=TI.state, values=data["state"])
+    query = _apply_array_filter(query, key=TI.pool, values=data["pool"])
+    query = _apply_array_filter(query, key=TI.queue, values=data["queue"])
+
+    query = query.join(
+        SlaMiss,
+        and_(
+            SlaMiss.dag_id == TI.dag_id,
+            SlaMiss.task_id == TI.task_id,
+            SlaMiss.execution_date == TI.execution_date,
+        ),
+        isouter=True,
+    ).add_entity(SlaMiss)
```

Comment:
@mik-laj I have dealt with the n +1 problem, but I am unsure if this query is the best way to go forward.

Reply:
@mik-laj is there a better approach that I should work on or is this fine?

Reply:
LGTM.